src/dmd/dmodule.d

```diff
@@ -535,6 +501,77 @@ extern (C++) final class Module : Package
         return m;
     }
 
+    /**
+     * Converts a chain of identifiers to the filename of the module
+     *
+     * Params:
+     *  packages = the names of the "parent" packages
+     *  ident = the name of the child package or module
+     *
+     * Returns:
+     *  the filename of the child package or module
+     */
+    static const(char)* getFilename(Identifiers* packages, Identifier ident)
+    {
+        const(char)[] filename = ident.toString();
+
+        if (packages == null || packages.dim == 0)
+            return filename.ptr;
+
+        OutBuffer buf;
+        OutBuffer dotmods;
+        auto ms = global.params.modFileAliasStrings;
+        const msdim = ms ? ms.dim : 0;
+
+        void checkModFileAlias(const(char)[] p)
+        {
+            /* Check and replace the contents of buf[] with
+            * an alias string from global.params.modFileAliasStrings[]
+            */
+            dotmods.writestring(p);
+        Lmain:
+            for (size_t j = msdim; j--;)
+            {
+                const m = (*ms)[j];
+                const q = strchr(m, '=');
+                assert(q);
+                if (dotmods.offset == q - m && memcmp(dotmods.peekString(), m, q - m) == 0)
+                {
+                    buf.reset();
+                    auto qlen = strlen(q + 1);
+                    if (qlen && (q[qlen] == '/' || q[qlen] == '\\'))
+                        --qlen;             // remove trailing separator
+                    buf.writestring(q[1 .. qlen + 1]);
+                    break Lmain;            // last matching entry in ms[] wins
+                }
+            }
+            dotmods.writeByte('.');
+        }
+
+        foreach (pid; *packages)
+        {
+            const p = pid.toString();
+            buf.writestring(p);
+            if (msdim)
+                checkModFileAlias(p);
+            version (Windows)
+            {
+                buf.writeByte('\\');
+            }
+            else
+            {
+                buf.writeByte('/');
+            }
+        }
+        buf.writestring(filename);
+        if (msdim)
+            checkModFileAlias(filename);
+        buf.writeByte(0);
+        filename = buf.extractData().toDString();
```

Comment:
Note the converstion `toDString()`.

Reply:
OutBuffer really needs an `extractSlice()` method, which can replace lines 569 and 570. This will return a `char[]`, which can be returned directly without diverting into `toDString()`.

Reply:
https://github.com/dlang/dmd/pull/9368